# Agents and Tools

Agents involve an LLM making decisions about which Actions to take, taking that Action, seeing an Observation, and repeating that until done.


When used correctly agents can be extremely powerful. In order to load agents, you should understand the following concepts:

- Tool: A function that performs a specific duty. This can be things like: Google Search, Database lookup, Python REPL, other chains.
- LLM: The language model powering the agent.
- Agent: The agent to use.


For example I have to travel from Dubai to Canada, I type this in ChatGPT

Give me two flight options from Dubai to Canada on September 1, 2024 | ChatGPT will not be able to answer because has knowledge till
September 2021

ChatGPT plus has Expedia Plugin, if we enable this plugin it will go to Expedia Plugin and will try to pull information about Flights & it will show the information

When we think about LLM. Many people think that it is just a knowledge engine, it has knowledge and it will try to give answer based on that knowledge but the knowledge is only limited to September 2021. The think that most people missout is that Large Lanaguage Model has a reasoning engine, and using that reasoning engine it can figure out when someone types this type of Question


Give me  two flight options from Dubai to Canada on September 1, 2024


As a human we go to Expedia as we have a reasoning engine in our brain.

LLM has a reasoning engine as well, so it will figure out the Source, Destination, Date and it will call Expedia Plugin and it will return response back.

#**How much is US GDP in 2024? plus 5**


# Agent will conenct with external tools and it will use LLM reasoning capabilities

# To access Google Search Results in Real Time we use serpapi

# serpapi and llm-math tool

In [1]:
!pip install google-search-results

  Preparing metadata (setup.py) ... done
  Created wheel for google-search-results: filename=google_search_results-2.4.2-py3-none-any.whl size=32009 sha256=3375f1d431d730358280c03ea1c860ed3d1a828e85fea0ca82aa625bf030ee03
  Stored in directory: /root/.cache/pip/wheels/6e/42/3e/aeb691b02cb7175ec70e2da04b5658d4739d2b41e5f73cd06f
Successfully built google-search-results


In [2]:
!pip install langchain

In [3]:
!pip install -U langchain-community

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 41.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 4.4 MB/s eta 0:00:00


In [4]:
%pip install -qU langchain-openai

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.2/61.2 kB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 418.7/418.7 kB 31.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 606.1/606.1 kB 48.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 71.6 MB/s eta 0:00:00


In [5]:
import os

In [8]:
os.environ["SERPAPI_API_KEY"]='2d703a1989128b0459c7372e8cf8a300e4a9b5c8714614f73f35fad63506b9eb'

In [6]:
from google.colab import userdata
OPENAI_API_KEY=userdata.get('OPENAI_API_KEY')

In [7]:
os.environ["OPENAI_API_KEY"]=OPENAI_API_KEY

In [9]:
from langchain.agents import AgentType,initialize_agent, load_tools

In [10]:
from langchain_openai import ChatOpenAI

In [11]:
llm = ChatOpenAI(
    model="gpt-3.5-turbo",
    temperature=0)

In [12]:
tool = load_tools(["serpapi","llm-math"],llm=llm)

In [13]:
agent1 = initialize_agent(tool, llm, agent= AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose= True)

<ipython-input-13-4a01bdb96f25>:1: LangChainDeprecationWarning: LangChain agents will continue to be supported, but it is recommended for new use cases to be built with LangGraph. LangGraph offers a more flexible and full-featured framework for building agents, including support for tool-calling, persistence of state, and human-in-the-loop workflows. For details, refer to the `LangGraph documentation <https://langchain-ai.github.io/langgraph/>`_ as well as guides for `Migrating from AgentExecutor <https://python.langchain.com/docs/how_to/migrate_agent/>`_ and LangGraph's `Pre-built ReAct agent <https://langchain-ai.github.io/langgraph/how-tos/create-react-agent/>`_.
  agent1 = initialize_agent(tool, llm, agent= AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose= True)


In [14]:
agent1.run("How much is US GDP in 2024 plus 5?")

<ipython-input-14-f878089be4d1>:1: LangChainDeprecationWarning: The method `Chain.run` was deprecated in langchain 0.1.0 and will be removed in 1.0. Use :meth:`~invoke` instead.
  agent1.run("How much is US GDP in 2024 plus 5?")




> Entering new AgentExecutor chain...
I need to calculate the US GDP in 2024 and then add 5 to it.
Action: Calculator
Action Input: US GDP in 2024
Observation: Answer: 21433000000000.0
Thought:Now I need to add 5 to the US GDP in 2024.
Action: Calculator
Action Input: 21433000000000.0 + 5
Observation: Answer: 21433000000005.0
Thought:I now know the final answer.
Final Answer: US GDP in 2024 plus 5 is $21,433,000,000,005.00.

> Finished chain.


'US GDP in 2024 plus 5 is $21,433,000,000,005.00.'

In [15]:
agent1.run("in which year indian team won its second world cup give me all the player name of the winning team??")



> Entering new AgentExecutor chain...
I need to search for the year when the Indian team won its second World Cup and then find the names of all the players in that winning team.
Action: Search
Action Input: "Indian cricket team second World Cup win year"
Observation: India has participated in every edition of the Cricket World Cup since its introduction. India has won it twice in 1983 and 2011, and is amongst the only three teams to have won it more than once.
Thought:I need to search for the names of all the players in the Indian cricket team that won the World Cup in 2011.
Action: Search
Action Input: "Indian cricket team 2011 World Cup winning players"
Observation: {'type': 'organic_result', 'title': 'India Squad - ICC Cricket World Cup 2011', 'source': 'ESPNcricinfo'}
Thought:I need to click on the source from ESPNcricinfo to find the names of the players in the Indian cricket team that won the 2011 World Cup.
Action: Search
Action Input: "India Squad - ICC Cricket World Cup 201

'The Indian cricket team won its second World Cup in 2011. The players in the winning team were MS Dhoni, Virender Sehwag, Sachin Tendulkar, Gautam Gambhir, Zaheer Khan, and others.'

In [17]:
agent1.run("How much is scope of curry powder and scope of rice flour and wheat market in India 2025?")



> Entering new AgentExecutor chain...
I need to search for information on the scope of curry powder, rice flour, and wheat market in India in 2025.
Action: Search
Action Input: "scope of curry powder market in India 2025", "scope of rice flour market in India 2025", "scope of wheat market in India 2025"
Observation: [' entity_type: related_questions.', 'The organic rice flour market in India is expected to reach a projected revenue of US$ 124.6 million by 2030. A compound annual growth rate of 7.6% is expected.', 'India is likely to produce a record 115.4 million metric tons of wheat in 2025, the farm ministry said on Monday, as higher state-set ...', 'Flour Market size is estimated to reach $198.7 billion by 2027, growing at a CAGR of 3.7% during the forecast period 2022-2027.', 'The flour, rice and malt market size has grown strongly in recent years. It will grow from $149.85 billion in 2024 to $159.59 billion in 2025 at a compound ...', 'The Rice Flour Market size was valued at US

ValueError: Got error from SerpAPI: Google hasn't returned any results for this query.

In [18]:
agent1.run("who won the india's 2024 general election?")



> Entering new AgentExecutor chain...
I need to search for the results of India's 2024 general election.
Action: Search
Action Input: "India 2024 general election winner"
Observation: ['See results about entity_type: video_universal.', 'General elections were held in India from 19 April to 1 June 2024 in seven phases, to elect all 543 members of the Lok Sabha. Votes were counted and the ...', "The results of India's general elections to constitute 18th Lok Sabha, held in April–June 2024 were announced on 4th and 5th June 2024.", 'General Election to Parliamentary Constituencies: Trends & Results June-2024 ; Bharatiya Janata Party - BJP, 240, 0 ; Indian National Congress - ...', "See results and maps from India's 2024 Lok Sabha elections.", 'Prime Minister Narendra Modi and his Bharatiya Janata Party (BJP) won the most seats for the third consecutive time, but have fallen short of a majority in the ...', 'Results of the Indian general elections were out on Tuesday with the National De

'Bharatiya Janata Party (BJP) led by Prime Minister Narendra Modi'

In [19]:
agent1.run("how much seat BJP won in india's general election in 2024 plus 10000?")



> Entering new AgentExecutor chain...
I need to search for the number of seats BJP won in India's general election in 2024 first.
Action: Search
Action Input: "BJP seats won in India general election 2024"
Observation: ['General elections were held in India from 19 April to 1 June 2024 in seven phases, to elect all 543 members of the Lok Sabha. Votes were counted and the ...', "The results of India's general elections to constitute 18th Lok Sabha, held in April–June 2024 were announced on 4th and 5th June 2024.", 'General Election to Parliamentary Constituencies: Trends & Results June-2024 ; Bharatiya Janata Party - BJP, 240, 0 ; Indian National Congress - ...', "Prime Minister Narendra Modi will serve a third term in office after India's election authority confirmed the National Democratic Alliance coalition headed ...", 'Prime Minister Narendra Modi and his Bharatiya Janata Party (BJP) won the most seats for the third consecutive time, but have fallen short of a majority in the ...

'10240'

In [20]:
!pip install wikipedia

  Preparing metadata (setup.py) ... done
  Created wheel for wikipedia: filename=wikipedia-1.4.0-py3-none-any.whl size=11679 sha256=afb191040a228cc81f14ad0d1f5fb1d2e3a90c0390f0c58e361a85181b5b292f
  Stored in directory: /root/.cache/pip/wheels/8f/ab/cb/45ccc40522d3a1c41e1d2ad53b8f33a62f394011ec38cd71c6
Successfully built wikipedia


In [21]:
tool = load_tools(["wikipedia","llm-math"],llm=llm)

In [22]:
agent = initialize_agent(
    tool,
    llm,
    agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,
    verbose=True,

)

In [23]:
agent.run("in which year first indian film released?")



> Entering new AgentExecutor chain...
I need to find out the year when the first Indian film was released.
Action: wikipedia
Action Input: First Indian film
Observation: Page: Cinema of India
Summary: The cinema of India, consisting of motion pictures made by the Indian film industry, has had a large effect on world cinema since the second half of the 20th century. Indian cinema is made up of various film industries, each focused on producing films in a specific language, such as Hindi, Telugu, Tamil, Malayalam, Kannada, Bengali, Marathi,Punjabi, Bhojpuri, Assamese, Odia and others.
Major centres of film production across the country include Mumbai, Hyderabad, Chennai, Kolkata, Kochi, Bengaluru, Bhubaneswar-Cuttack, and Guwahati. For a number of years, the Indian film industry has ranked first in the world in terms of annual film output. In 2022, Indian cinema earned ₹15,000 crore ($1.9 billion) at the box-office. Ramoji Film City located in Hyderabad is certified by the Guinness Wor

'The first Indian film was released in 1913.'

In [ ]:
agent.run("give me a detail analysis of indian election 2024")



> Entering new AgentExecutor chain...
I should use Wikipedia to find information about the Indian election in 2024.
Action: wikipedia
Action Input: Indian election 2024
Observation: Page: 2024 Indian general election
Summary: General elections were held in India from 19 April to 1 June 2024 in seven phases, to elect all 543 members of the Lok Sabha. Votes were counted and the result was declared on 4 June to form the 18th Lok Sabha. On 7 June 2024, Modi confirmed the support of 293 MPs to Droupadi Murmu, the president of India. This marked Modi's third term as prime minister and his first time heading a coalition government, with the Telugu Desam Party of Andhra Pradesh and Janata Dal (United) of Bihar emerging as two main allies.
More than 968 million people out of a population of 1.4 billion people were eligible to vote, equivalent to 70 percent of the total population. 642 million voters participated in the election and 312 million of them were women, making it the highest ever pa

'The 2024 Indian general election was held in seven phases from 19 April to 1 June 2024 to elect all 543 members of the Lok Sabha. Narendra Modi secured his third term as prime minister, leading a coalition government with the support of 293 MPs. The election saw a record number of eligible voters and high participation, with the BJP losing its majority in the Lok Sabha.'

The 2024 Indian general election was held in seven phases from 19 April to 1 June 2024 to elect all 543 members of the Lok Sabha. Narendra Modi secured his third term as prime minister, leading a coalition government with the support of 293 MPs. The election saw a record number of eligible voters and high participation, with the BJP losing its majority in the Lok Sabha.